---
title: "Dicaeum SNP filtering"
author: "Devon DeRaad"
date: '2022-11-29'
output: html_document
---

```{r}
library(SNPfiltR)
library(vcfR)
#read in vcf as vcfR
v<- read.vcfR("~/Desktop/phil.dicaeum/n8.vcf")
#make popmap
pops<-read.csv("~/Desktop/phil.dicaeum/dicaeum.sampling.csv")
pops<-pops[pops$ID %in% colnames(v@gt),]
popmap<-pops[,c(1,4)]

#remove obviously contaminated sample from looking at the preliminary splitstree
v<-v[,colnames(v@gt)!= "D_hypoleucum_28663"]
#fix popmap
pops<-pops[pops$ID %in% colnames(v@gt),]
popmap<-pops[,c(1,4)]
```

```{r}
#visualize distributions
hard_filter(v)
#hard filter to minimum depth of 5, and minimum genotype quality of 30
v<-hard_filter(v, depth = 5, gq = 35)
```

```{r}
#execute allele balance filter
v<-filter_allele_balance(v)
```

```{r}
#visualize and pick appropriate max depth cutoff
max_depth(v)
#filter for max depth
max_depth(v, maxdepth = 150)
#remove invariant sites generated by filtering
v<-min_mac(v, min.mac = 1)
#check vcfR to see how many SNPs we have left
v
```

```{r}
#investigate patterns of missingness per sample
missing_by_sample(v)
#investigate patterns of missingness per SNP
missing_by_snp(v)
#implement per sample completeness filter
v<-missing_by_sample(v, cutoff = .785)
#remove invariant sites generated by dropping individuals
v<-min_mac(v, min.mac = 1)
#fix popmap
pops<-pops[pops$ID %in% colnames(v@gt),]
popmap<-popmap[popmap$ID %in% colnames(v@gt),]
colnames(popmap)<-c('id','pop')
```


```{r}
missing_by_snp(v)
#reorder popmap to match vcf, this is a bug, you shouldn't have to do this
popmaps<-popmap[order(match(popmap$id,colnames(v@gt)[-1])),]
miss<-assess_missing_data_tsne(v, popmap=popmaps, threshold=c(.85,.9,.95), clustering = FALSE)
#missing data doesn't seem to be driving clustering regardless of threhsold
#implement missing data per SNP filter
v<-missing_by_snp(v, cutoff=.85)
```


```{r}
#investigate clustering patterns with a minor allele cutoff
v.mac<-min_mac(v, min.mac = 2)
v.mac
assess_missing_data_tsne(v.mac, popmap=popmaps, clustering = FALSE)
#seeing potentially spurious groupings with singletons removed, we will ignore for now
```

```{r}
#plot depth per snp and per sample
dp <- extract.gt(v, element = "DP", as.numeric=TRUE)
heatmap.bp(dp, rlabels = FALSE)
#plot genotype quality per snp and per sample
gq <- extract.gt(v, element = "GQ", as.numeric=TRUE)
heatmap.bp(gq, rlabels = FALSE)
```


```{r}
#write out vcf with all SNPs
#vcfR::write.vcf(v, "~/Desktop/phil.dicaeum/filtered.85.vcf.gz")

#linkage filter vcf to thin SNPs to one per 500bp
v.thin<-distance_thin(v, min.distance = 500)

#write out thinned vcf
#vcfR::write.vcf(v.thin, "~/Desktop/phil.dicaeum/filtered.85.unlinked.vcf.gz")
```

