### Demultiplexing
* I used the script called 'demultiplex.sh' to demultiplex the individual sample files based on barcode using the Stacks function [process_radtags](https://catchenlab.life.illinois.edu/stacks/manual/#procrad). ~160 samples shared across projects were put on a single NextSeq2000 run, outputting 400M single end 100bp reads. These samples were dual indexed, so that we could use only ~80 unique barcodes (2*80 = 160 unique combinations) to identify the reads from each sample. The reads were returned to us from the KU genome sequencing core already demultiplexed into two separate files, each containing ~80 single indexed samples. I used the files 'plate1.barcodes.txt' and 'plate2.barcodes.txt' to demultiplex each sample's reads into a unique fastq file that corresponds to that sample. This had to be done separately because the two sets of barcodes overlap, so doing them together will artificially combine reads from multiple samples and mess everything up.

### Stacks parameter optimization
* I used the scripts 'optimize.m.sh', 'optimize.bigM.sh', and 'optimize.n.sh' to perform iterative parameter optimization of the key parameters for *de novo* RAD locus assembly, following the protocol outlined by [Paris et al. 2017](https://besjournals.onlinelibrary.wiley.com/doi/10.1111/2041-210X.12775). The entire optimization protocol including corresponding visualizations is available for viewing at: [https://devonderaad.github.io/phil.dicaeum.rad/sequence.data.to.snps/optimize.denovo.diacaeum.assembly.html](https://devonderaad.github.io/phil.dicaeum.rad/sequence.data.to.snps/optimize.denovo.diacaeum.assembly.html).
